R/load_weights.R

```diff
@@ -0,0 +1,157 @@
+
+  for (qn in query_names) {
+    kn <- stringr::str_replace(string = qn,
+                               pattern = trigger_pattern,
+                               replacement = "key.")
+    vn <- stringr::str_replace(string = qn,
+                               pattern = trigger_pattern,
+                               replacement = "value.")
+    ipn <- stringr::str_replace(string = qn,
+                                pattern = trigger_pattern,
+                                replacement = "in_proj_")
+    combined <- torch::torch_cat(list(state_dict[[qn]],
+                                      state_dict[[kn]],
+                                      state_dict[[vn]]))
+    state_dict[[ipn]] <- combined
+    state_dict[[qn]] <- NULL
+    state_dict[[kn]] <- NULL
+    state_dict[[vn]] <- NULL
+  }
+  return(state_dict)
+}
+
+#' Clean up Parameter Names
+#'
+#' There are some hard-to-avoid differences between the variable names in BERT
+#' models constructed using this package and the standard variable names used in
+#' the Huggingface saved weights. This function changes the names from the
+#' Huggingface saves to match package usage.
+#'
+#' @param state_dict A state_dict of pretrained weights, probably loaded from a
+#'  file.
+#'
+#' @return The state_dict with the names normalized.
+#' @keywords internal
+.rename_state_dict_variables <- function(state_dict) {
+  rep_rules <- variable_names_replacement_rules
+  w_names <- names(state_dict)
+  w_names <- stringr::str_replace_all(string = w_names,
+                                      pattern = stringr::fixed(rep_rules))
+  names(state_dict) <- w_names
+  return(state_dict)
+}
+
+
+#' Load Pretrained Weights into a BERT Model
+#'
+#' There are some hard-to-avoid differences between the variable names in BERT
+#' models constructed using this package and the standard variable names used in
+#' the Huggingface saved weights. This function changes the names from the
+#' Huggingface saves to match package usage.
+#'
+#' @param model a BERT-type model, constructed using `model_bert`.
+#' @param model_name Character; which flavor of BERT to use. Must be compatible
+#'   with `model`!
+#'
+#' @return The number of model parameters updated. (This is to enable error
+#'   checks; the function is called for side effects.)
+#' @keywords internal
+.load_weights <- function(model, model_name = "bert_base_uncased") {
+  # once the cache is set up, this should usually just be fetching from there
+  sd <- .download_weights(model_name = model_name)
+
+  my_sd <- model$state_dict()
+  my_weight_names <- names(my_sd)
+  saved_weight_names <- names(sd)
+  names_in_common <- intersect(my_weight_names, saved_weight_names)
+  if (length(names_in_common) > 0) {
+    my_sd[names_in_common] <- sd[names_in_common]
+  } else {
+    warning("No matching weight names found.") # nocov
+  }
+  model$load_state_dict(my_sd)
+  return(length(names_in_common)) #maybe? This function is for side effects.
```

Comment:
I think this is a fine thing to do.
```suggestion
  return(length(names_in_common)) # This function is for side effects.
```